analyzers/src/SonarAnalyzer.Common/Rules/UseIFormatProviderForParsingDateAndTimeBase.cs

```diff
@@ -0,0 +1,79 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Globalization;
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UseIFormatProviderForParsingDateAndTimeBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S6580";
+
+    private static readonly string[] parseMethodNames = new[]
+    {
+        nameof(DateTime.Parse),
+        nameof(DateTime.ParseExact),
+        nameof(DateTime.TryParse),
+        nameof(DateTime.TryParseExact)
+    };
+    private static readonly string[] nonDeterministicFormatProviders = new[]
+    {
+        nameof(CultureInfo.CurrentCulture),
+        nameof(CultureInfo.CurrentUICulture),
+        nameof(CultureInfo.DefaultThreadCurrentCulture),
+        nameof(CultureInfo.DefaultThreadCurrentUICulture)
+    };
+    private static readonly KnownType[] temporalTypes = new[]
+    {
+        KnownType.System_DateOnly,
+        KnownType.System_DateTime,
+        KnownType.System_DateTimeOffset,
+        KnownType.System_TimeOnly,
+        KnownType.System_TimeSpan
+    };
+
+    protected override string MessageFormat => "Pass an 'IFormatProvider' to the '{0}' method.";
```

Comment:
I'm not sure about the message. Any better ideas? The original was `Always use DateTime.Parse overloads with an "IFormatProvider" parameter`. I didn't like it, because using the overloaded method isn't enough, you can still pass `null` or `CultureInfo.CurrentCulture`.